dialects/direct/passes/ds2cps.cpp

```diff
@@ -0,0 +1,266 @@
+#include "dialects/direct/passes/ds2cps.h"
+
+#include <iostream>
+
+#include <thorin/lam.h>
+
+#include "dialects/direct/direct.h"
+
+// #define verbose_rewrite
+
+namespace thorin::direct {
+
+/// called by the pass manager when the pass is run
+/// for every lambda in the code
+/// but we need some manual invokation for new lambdas that need urgent care
+void DS2CPS::enter() {
+    Lam* lam = curr_nom();
+    rewrite_lam(lam);
+}
+
+bool isDS(Lam* lam) {
+    auto ty       = lam->type()->as<Pi>();
+    auto codom_ty = ty->codom();
+    return !(codom_ty->isa<Bot>());
+}
+
+/// switches context to new lambda
+/// replaces the body
+/// note: if the lambda contains a ds call, the body will be replaced
+///   with a cps call to the previously ds function and the original
+///   body will be altered and stored in a new continuation
+void DS2CPS::rewrite_lam(Lam* lam) {
+    if (auto i = rewritten_.find(lam); i != rewritten_.end()) return;
+
+    Lam* prev     = currentLambda;
+    currentLambda = lam;
+
```

Comment:
just for the sake of consistency:
`currentLambda` -> `curr_lam` or maybe just `curr`. It's also just `prev` so I guess `curr` is just fine :)